### ISSUE TITLE
Scaling a pytorch tensor by a python float radically changes normalization behavior

### ISSUE THREAD

**Issue by capybaralet** [pytorch/pytorch#17260]
## 🐛 Bug
See title

## To Reproduce

Run this code:

```
import torch as T

W = T.randn(1, 3).expand(2, 3)
WW = 1.0 * W 
print((WW == W).sum())
W /= (((W**2).sum(1))**.5).unsqueeze(1)
WW /= (((WW**2).sum(1))**.5).unsqueeze(1)
print((WW == W).sum())

# The effect is not small!
print ((WW**2).sum())
print ((W**2).sum())

# This works
W = T.randn(1, 3).expand(2, 3)
WW = 1.0 * W 
print((WW == W).sum())
W_ = (((W**2).sum(1))**.5).unsqueeze(1)
W = W / W_
WW_ = (((WW**2).sum(1))**.5).unsqueeze(1)
WW = WW / WW_
print((WW == W).sum())

```

## Expected behavior

WW and W should remain the same after the normalization.

## Environment

Collecting environment information...
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Ubuntu 18.10
GCC version: (Ubuntu 8.2.0-7ubuntu1) 8.2.0
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] numpydoc==0.8.0
[pip] torch==1.0.0
[pip] torchvision==0.2.1
[conda] blas                      1.0                         mkl                                                                                                                                                                                            
[conda] mkl                       2019.1                      144                                                                                                                                                                                            
[conda] mkl-service               1.1.2            py37he904b0f_5                                                                                                                                                                                            
[conda] mkl_fft                   1.0.6            py37hd81dba3_0                                                                                                                                                                                            
[conda] mkl_random                1.0.2            py37hd81dba3_0                                                                                                                                                                                            
[conda] pytorch-cpu               1.0.0               py3.7_cpu_1    pytorch                                                                                                                                                                                 
[conda] torchvision-cpu           0.2.1                      py_2    pytorch

**Comment by capybaralet** [comment]
I guess this is the same issue as https://github.com/pytorch/pytorch/issues/7731****

**Comment by capybaralet** [comment]
But I still don't understand why scaling the tensor would change the behavior of the in-place operations....

**Comment by colesbury** [comment]
See #17185 and #957.

In-place operations on tensors with with overlapping storage results in undefined behavior.

You can see this behavior with a simplified version of your above example. In-place division results in a different result than out-of-place division because there's only three storage locations (not six), so some storage locations get divided twice.

```python
>>> import torch
>>> W = torch.tensor([[0.0, 1.0, 2.0]]).expand(2, 3)
>>> print(W)
tensor([[0., 1., 2.],
        [0., 1., 2.]])
>>> print(W / 2)
tensor([[0.0000, 0.5000, 1.0000],
        [0.0000, 0.5000, 1.0000]])
>>> W /= 2
>>> print(W)
tensor([[0.0000, 0.2500, 0.5000],
        [0.0000, 0.2500, 0.5000]])
```

**Comment by colesbury** [comment]
By the way, I should say this is confusing behavior and we are working on fixing it.
